src/createQueryGroups.ts

```diff
@@ -25,74 +37,109 @@ export type CreateQueryGroups = {
 }
 
 export function createQueryGroups(options?: QueryGroupOptions) {
-  const actions = new Map<QueryAction, number>()
-  const actionGroups = new Map<number, Set<QueryGroupKey>>()
-  const groups = new Map<QueryGroupKey, QueryGroup>()
-
-  function getActionKey(action: QueryAction): number {
-    if(!actions.has(action)) {
-      actions.set(action, createActionId())
-    }
-
-    return actions.get(action)!
-  }
+  const queries = createIndexedCollection([] as QueryData[], ['queryGroupId', 'actionId'])
+  const tags = createIndexedCollection([] as TagData[], ['queryGroupId', 'key'])
 
-  function getQueryGroupKey(actionKey: number, args: Parameters<QueryAction>): QueryGroupKey {
-    return `${actionKey}-${JSON.stringify(args)}`
+  function createQueryGroupId(actionId: number, args: Parameters<QueryAction>): QueryGroupId {
+    return `${actionId}-${JSON.stringify(args)}`
```

Comment:
Off topic: Since both action ids and group ids are idempotent we should abstract them both in the same way. 

Preferred
```
getActionId
getQueryGroupId
```
or
```
createActionId
createQueryGroupId
```

The query group id utility could be its own service/utility as well. Its not specific to the groups instance.

Reply:
This is something we could/should just PR separately if you agree.